File: src/main/java/jp/kentan/minecraft/neko_event/config/provider/SignConfigProvider.java
package jp.kentan.minecraft.neko_event.config.provider;

import jp.kentan.minecraft.neko_event.config.ConfigManager;
import jp.kentan.minecraft.neko_event.util.Log;
import jp.kentan.minecraft.neko_event.util.NekoUtil;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.block.Sign;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;

import java.io.File;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;


public class SignConfigProvider {

    private static File sConfigFile;

    private static Map<Location, Metadata> sSignMetadataCacheMap = new HashMap<>();

    public static void setup(File dataFolder){
        sConfigFile = new File(dataFolder + File.separator + "sign.yml");

        ConfigManager.createIfNeed(sConfigFile);

        load();
    }

    private static void load(){
        try {
            FileConfiguration config = new YamlConfiguration();
            config.load(sConfigFile);

            if(!config.isConfigurationSection("Sign")){
                Log.warn("Sign section not found.");
                return;
            }

            final Map<Location, Metadata> signMetadataMap = new HashMap<>();

            final Set<String> indexSet = config.getConfigurationSection("Sign").getKeys(false);

            indexSet.forEach(index -> {
                final String path = "Sign." + index;

                Location location = new Location(
                        Bukkit.getWorld(config.getString(path + ".Location.world")),
                        config.getDouble(path + ".Location.x"),
                        config.getDouble(path + ".Location.y"),
                        config.getDouble(path + ".Location.z")
                );

                if(!(location.getBlock().getState() instanceof Sign)){
                    config.set(path, null);
                    Log.warn("看板" + NekoUtil.toString(location) + "が見つかりませんでした.");
                    return;
                }

                Map<String, Object> metadataMap = new HashMap<>();

                final Set<String> metadataSet = config.getConfigurationSection(path + ".Metadata").getKeys(false);
                final String metadataPath = path + ".Metadata.";
                metadataSet.forEach(key -> metadataMap.put(key, config.get(metadataPath + key)));

                if(signMetadataMap.containsKey(location)){
                    Log.error("看板ﾒﾀﾃﾞｰﾀが重複しています. (index:" + index + ")");
                    return;
                }

                signMetadataMap.put(location, new Metadata(index, metadataMap));
            });

            config.save(sConfigFile);

            sSignMetadataCacheMap.clear();
            sSignMetadataCacheMap.putAll(signMetadataMap);

            Log.info(sSignMetadataCacheMap.size() + "件の看板ﾒﾀﾃﾞｰﾀを読み込みました.");
        } catch (Exception e) {
            e.printStackTrace();
            Log.error(e.getMessage());
        }
    }

    public static boolean hasMetadata(Location location, String key){
        return sSignMetadataCacheMap.containsKey(location) && sSignMetadataCacheMap.get(location).containsKey(key);
    }

    public static Object getMetadata(Location location, String key){
        return sSignMetadataCacheMap.get(location).get(key);
    }

    public static boolean saveMetadata(Location location, Map<String, Object> metadataMap) {
        final String index, path;

        try {
            FileConfiguration config = new YamlConfiguration();
            config.load(sConfigFile);

            if(sSignMetadataCacheMap.containsKey(location)){
                save(new HashMap<String, Object>() {
                    {
                        put("Sign." + sSignMetadataCacheMap.get(location).getIndex(), null);
                    }
                });
            }

            String signIndex = null;
            for(int i = 0; i < 500; ++i){
                if(!config.isConfigurationSection("Sign." + i)){
                    signIndex = Integer.toString(i);
                    break;
                }
            }

            if(signIndex == null){
                Log.error("看板が保存上限に到達しています.");
                return false;
            }

            index = signIndex;
            path = "Sign." + signIndex;
        } catch (Exception e) {
            e.printStackTrace();
            Log.error(e.getMessage());
            return false;
        }

        if (save(new LinkedHashMap<String, Object>(){
            {
                put(path + ".Location.world", location.getWorld().getName());
                put(path + ".Location.x", location.getX());
                put(path + ".Location.y", location.getY());
                put(path + ".Location.z", location.getZ());

                final String metadataPath = path + ".Metadata.";

                metadataMap.forEach((key, val) -> put(metadataPath + key, val));
            }
        })){
            sSignMetadataCacheMap.put(location, new Metadata(index, metadataMap));
            return true;
        }else{
            return false;
        }
    }

    private static boolean save(Map<String, Object> dataList) {
        try {
            FileConfiguration config = new YamlConfiguration();
            config.load(sConfigFile);

            dataList.forEach(config::set);

            config.save(sConfigFile);
        } catch (Exception e) {
            e.printStackTrace();
            Log.error(e.getMessage());

            return false;
        }
        return true;
    }


    private static class Metadata{
        private String mIndex;
        private Map<String, Object> mMetadataMap = new HashMap<>();

        Metadata(String index, Map<String, Object> metadataMap){
            mIndex = index;
            mMetadataMap.putAll(metadataMap);
        }

        Object get(String key){
            return mMetadataMap.get(key);
        }

        String getIndex(){
            return mIndex;
        }

        boolean containsKey(String key){
            return mMetadataMap.containsKey(key);
        }
    }
}

File: src/main/java/jp/kentan/minecraft/nekoevent/config/provider/SignConfigProvider.kt
package jp.kentan.minecraft.nekoevent.config.provider

import jp.kentan.minecraft.nekoevent.util.Log
import jp.kentan.minecraft.nekoevent.util.formatString
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.block.Sign
import org.bukkit.configuration.file.YamlConfiguration
import java.io.File
import java.util.LinkedHashMap

class SignConfigProvider(dataFolder: File) : BaseConfigProvider(dataFolder, "sign.yml") {

    private val signMetadataMap = mutableMapOf<Location, Metadata>()

    init {
        load()
    }

    fun load() {
        try {
            val config = YamlConfiguration.loadConfiguration(configFile)

            if (!config.isConfigurationSection("Sign")) {
                Log.warn("Signセクションが見つかりません.")
                return
            }

            val indexSet = config.getConfigurationSection("Sign").getKeys(false)

            val metadataMap = indexSet.mapNotNull { strIndex ->
                val path = "Sign.$strIndex"

                val location = Location(
                        Bukkit.getWorld(config.getString("$path.Location.world")),
                        config.getDouble("$path.Location.x"),
                        config.getDouble("$path.Location.y"),
                        config.getDouble("$path.Location.z")
                )

                //ワールドが見つからない場合は消去せずスキップ
                if (location.world == null) {
                    Log.warn("看板($strIndex)が見つかりませんでした.")
                    return@mapNotNull null
                }

                // 看板がない場合は消去
                if (location.block.state !is Sign) {
                    config.set(path, null)
                    Log.warn("看板${location.formatString()}が見つかりませんでした.")
                    return@mapNotNull null
                }

                val index = strIndex.toIntOrNull() ?: let {
                    Log.warn("看板($strIndex)のindexを整数に変換できません.")
                    return@mapNotNull null
                }


                val metadataSet = config.getConfigurationSection("$path.Metadata").getKeys(false)
                val metadataPath = "$path.Metadata."
                val metadataMap = metadataSet.associate { it to config.get(metadataPath + it) }.toMutableMap()

                return@mapNotNull location to Metadata(index, metadataMap)
            }.toMap()

            config.save(configFile)

            signMetadataMap.clear()
            signMetadataMap.putAll(metadataMap)

            Log.info("${signMetadataMap.size}件の看板ﾒﾀﾃﾞｰﾀを読み込みました.")
        } catch (e: Exception) {
            e.printStackTrace()
            Log.error(e)
        }
    }

    fun save(location: Location, dataMap: Map<String, Any>): Boolean {
        try {
            val config = YamlConfiguration.loadConfiguration(configFile)

            // クリーニング
            signMetadataMap[location]?.apply {
                super.save(mapOf("Sign.${this.index}" to null))
            }

            // 利用可能なindexを探す
            val index: Int = kotlin.run {
                var count = 0
                var prev = 0
                config.getConfigurationSection("Sign").getKeys(false)
                        .mapNotNull { it.toInt() }
                        .sorted()
                        .forEach {
                            count = it

                            if (count - prev > 1) { return@run prev + 1 }

                            prev = count
                        }

                return@run count + 1
            }

            val path = "Sign.$index"

            val metadataMap = LinkedHashMap<String, Any>().apply {
                put("$path.Location.world", location.world.name)
                put("$path.Location.x", location.x)
                put("$path.Location.y", location.y)
                put("$path.Location.z", location.z)

                val metadataPath = "$path.Metadata."

                dataMap.forEach { key, data -> put(metadataPath + key, data) }
            }

            if (super.save(metadataMap)) {
                signMetadataMap[location] = SignConfigProvider.Metadata(index, dataMap)
            }
        } catch (e: Exception) {
            e.printStackTrace()
            Log.error(e)
            return false
        }

        return true
    }

    fun getMetadata(location: Location, key: String): Any? {
        return signMetadataMap[location]?.get(key) ?: let {
            Log.error("看板" + location.formatString() + "にﾒﾀﾃﾞｰﾀ($key)が不足しています.")
            return null
        }
    }

    private class Metadata(
            val index: Int,
            val metadataMap: Map<String, Any>)
    {
        fun get(key: String) = metadataMap[key]
    }
}